### PR TITLE
Updated documentation for JNLP slaves for newer Jenkins versions

### PR DESCRIPTION
**Pull request by MadsNielsen**
Found out while creating a demo, that recent changes in remoting requires a new configuration for JNLP slaves. 

So i updated the documentation with the new tidbits.

### CODE REVIEW

File: demos/build_agents/README.md [hunk 1]
```diff
@@ -32,3 +34,41 @@ jenkins:
...
+  (...)
+  nodes:
+    - dumb:
+        labelString: "linux docker test"
+        mode: NORMAL
+        name: "utility-node"
+        remoteFS: "/home/user1"
+        launcher:
+          jnlp:
+            workDirSettings:
```
Comment (jglick) [inline demos/build_agents/README.md]: As in #168, I would hope this is optional.

File: demos/build_agents/README.md [hunk 2]
```diff
@@ -4,6 +4,8 @@ Build agents configuration belongs (currently) under `jenkins` root element
 
 ## sample configuration
 
+### Jenkins versions < 2.68
```
[the remark this answers is not in the document]
Comment (jglick) [reply demos/build_agents/README.md]: > do we have a policy about this

@kohsuke has not formally said anything, and if you are maintaining the plugin you set the policy, but [read this](https://github.com/jenkinsci/jep/blob/master/jep/300/README.adoc#automatically-updated-distribution) if you have not already.

**Comment by oleg-nenashev** [comment]
The plugin needs workDir settings as a mandatory dependency, because the field is marked as Nonnull here: https://github.com/jenkinsci/jenkins/blob/393680beb5d64628bdf8d9e5bf8daf4cab2191ad/core/src/main/java/hudson/slaves/JNLPLauncher.java#L69-L77 . 

There is a default constructor which does not require working directory, but IIUC this plugin is unable to discover that constructor.

It may be a serious compatibility issue for the CasC design - many plugins evolve by creating new constructors and deprecating existing ones. In such case the user's configuration YAML will be broken during update.

@MadsNielsen @ewelinawilkosz @ndeloof My proposal would be to see whether it is possible to update the plugin to heck for deprecated public constructor in order to retain compatibility in such case

**Comment by ndeloof** [comment]
config-as-code assumes databound constructor do cleanly reflect the component requirements. 
by design, update to core|plugins do change the data model and as such can require updates to yaml file. There's no such thing like a yaml backward compatibility expected here.

**Comment by oleg-nenashev** [comment]
OK 🤷‍♂️

**Comment by oleg-nenashev** [comment]
Obviously, we could modify Jenkins Core in this particular case.

**Comment by ndeloof** [comment]
I also expect plugin developer could take more care not changing DataBoundConstructor but for newly introduced required parameters. They can rely on DataBoundSetters to introduce optional parameters. AFAICT changing constructor has been used a lot in the past because we missed this option.

**Comment by oleg-nenashev** [comment]
`@DataBoundSetter` still has a slight design flaw. Whatever you move to  setter, becomes a part of public API, and you cannot make the field `final`. So one either has to handle non-constructor modifications or somehow restrict the field

**Comment by ndeloof** [comment]
> Whatever you move to setter, becomes a part of public API
Just like a public constructor parameter and associated getter for databinding, right ?

> and you cannot make the field final
As long as you don't explose it as a mutable data structure on getter, that shouldn't be an issue. Same issue applies in many places with data migration from `readResolve` with some `/* mostly final */` comment in code.

**Comment by oleg-nenashev** [comment]
IMHO we need https://github.com/jenkinsci/configuration-as-code-plugin/pull/234 instead, but I am fine with the merge if it helps to move forward

**Comment by ewelinawilkosz** [comment]
I'm just trying to clean up, haven't got to #234 yet

**Comment by MadsNielsen** [comment]
I'm fine with not merging this, if we can merge #234. I like @oleg-nenashev elgant solution. It's glue code for the first period, until people start writing proper plugins with correct use of @DataBound.

**Comment by oleg-nenashev** [comment]
@MadsNielsen It can be closed once #234 is released

**Comment by ewelinawilkosz** [comment]
#234 merged, closing this one as decided above

**Pull request closed without merging**